### PR TITLE
Remove deprecated method Series.set_index.

### LINKED ISSUE #9541: [DEPRECATION] Remove Series.set_index
pandas supports `DataFrame.set_index` to convert specified columns of a DataFrame into its index. To actually set a separate index object as the index of a DataFrame or Series, however, a user must set the `index` property (e.g. `df.index = RangeIndex(10)`). `Series.set_index` as implemented in cuDF (as an alias for setting the property) confuses the issue. The method was deprecated in #9529 for 21.12 and we will remove it in 22.02.

### PR DESCRIPTION
**Pull request by bdice**
This PR removes the deprecated method `Series.set_index`. Resolves #9541, follows up on #9529.

### CODE REVIEW

File: python/cudf/cudf/core/series.py
Comment (bdice) [inline python/cudf/cudf/core/series.py:3614]: I'd like a specific review of this change from @vyasr or @shwina when you have time. I removed the internal use of `set_index`. Does this change use the factory methods / copy semantics properly?

**Review by cwharris — APPROVED** [review]

File: python/cudf/cudf/core/series.py
Comment (vyasr) [reply python/cudf/cudf/core/series.py:3614]: What you're doing here should have the right semantics but is making more copies than necessary. `Series._from_data` does not make copies, it just assigns attributes by reference, which is why you need to do the `self.copy` in the first line. However, we pass `self.index` rather than `out.index` to `_from_data`, which is acceptable because indices in pandas are immutable. That being the case, you could make this more efficient by only copying `self._data` (`out_data  = self._data.copy(deep=copy)`) and passing that as the first argument to `_from_data`, avoiding the unnecessary copy of the index. I would be fine making that change as part of this PR, since you're already removing (additional) unnecessary copies from the old version.

It's worth noting that `rename` is not perfectly matching pandas, which accepts [a bunch of extra arguments](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.rename.html) and basically ignores them. In particular, the `copy` parameter is actually ignored by pandas:
```
>>> s = pd.Series([1, 2, 3])
>>> s1 = s.rename('a', copy=False)
>>> s1[0] = 10
>>> s
0    1
1    2
2    3
dtype: int64
>>> s1
0    10
1     2
2     3
Name: a, dtype: int64
```
Of course, we care more about memory constraints on the GPU than on the host, so maybe we're OK with this discrepancy. I'll defer to @shwina on how that should behave. Regardless of whether or not we decide to change it, any changes relating to the `rename` API are out of scope for this PR and we should make an issue documenting what we want to do going forward. Hopefully this comment is sufficient documentation for that.

Comment (shwina) [reply python/cudf/cudf/core/series.py:3614]: We should not create a copy unless `copy=True`.

Comment (bdice) [reply python/cudf/cudf/core/series.py]: Thanks to both of you -- this is resolved in e4fa9ddc80e138fe984cf3a45896027c1cc78db3.

**Comment by bdice** [comment]
@vyasr @shwina I'm re-requesting your review after applying your suggestions. I think I have everything fixed now, but I'll wait for a second approval before merging.

**Comment by bdice** [comment]
rerun tests

**Comment by vyasr** [comment]
rerun tests

File: python/cudf/cudf/core/multiindex.py
Comment (vyasr) [inline python/cudf/cudf/core/multiindex.py:781]: At this point it is safe to assume that index is always an index object, so I don't think you need this branching. 
```suggestion
            result.index = index
```

**Review by vyasr — CHANGES_REQUESTED** [review]
@bdice my apologies for missing this earlier. This PR is very small, but it's also very low priority, so I proposed a change that I might not have if I thought this was critical to get into 22.02. If you can make the change soon I think we can still get it into the release, but I also don't think it's a huge deal if it slips. Sorry for not seeing this sooner!

**Comment by vyasr** [comment]
@isVoid do we need to resolve the notebooks question before merging this?

**Comment by bdice** [comment]
@vyasr I am going to fix the notebooks and retarget 22.04. This has been on my list of things to get done but other 22.02 priorities superseded it. @isVoid Great catch with the notebooks!

**Comment by isVoid** [comment]
> @isVoid do we need to resolve the notebooks question before merging this?

No, other repos burndown starts later to cuDF. I don't think changes in notebooks should block this. @vyasr

File: python/cudf/cudf/core/multiindex.py
Comment (bdice) [reply python/cudf/cudf/core/multiindex.py:781]: @vyasr This is not branching on the type of `index`. It's branching on the type of `result`. If `result` is a `Series`, we set its `index` attribute. If `result` is a `DataFrame`, we call `DataFrame.set_index(index)`. I don't think `DataFrame.set_index` does the same thing as setting its `index` attribute.

Comment (bdice) [reply python/cudf/cudf/core/multiindex.py:781]: I tested this change locally and I see that `test_memory_usage_multi` fails with this change. It gets a size of 368 instead of 360. I take this to mean that there is some difference between calling `DataFrame.set_index` and setting the `DataFrame.index` attribute.

Comment (vyasr) [reply python/cudf/cudf/core/multiindex.py:781]: I chatted offline with @bdice about this. The failing test indicates to me that there is indeed a difference between the two, but not one that should be present. Fixing that is out of scope for this PR, so I've asked him to add an appropriate comment here and we can move forward with the code as is for now.

**Review by vyasr — DISMISSED** [review]
Approving pending the addition of the comment that I requested offline.

File: python/cudf/cudf/core/multiindex.py
Comment (bdice) [reply python/cudf/cudf/core/multiindex.py:790]: Comment added in 9ca4b4cf23.

**Comment by bdice** [comment]
I messaged @taureandyernv about updating the notebooks. The change I would suggest is to set the `index` attribute of the `Series` like `column.index = data`, instead of calling `new_column = column.set_index(data)`. We will also want to change the text in the cell right above it:

> We can create a new column with a different index by using the `set_index` method.

File: python/cudf/cudf/core/multiindex.py
Comment (vyasr) [reply python/cudf/cudf/core/multiindex.py:790]: @bdice sorry for turning this around again. I just tried to make this change locally using your branch and I find the failure to be extremely flaky. I observe the failure both with and without this change, at random intervals. You can use pytest-repeat with `pytest --count=100 python/cudf/cudf/tests/test_dataframe.py::test_memory_usage_multi` to hopefully see this for yourself. The failure is consistently that the computed memory usage underestimates the expected by 8 bytes (`368 != 360`) on a MultiIndex with 10 rows. I am investigating this further because I'm curious what the cause of this discrepancy is, but I think it means that you can safely make this change (and remove the TODO) in this PR.

Comment (vyasr) [reply python/cudf/cudf/core/multiindex.py:790]: The memory test issue is resolved in #10114. This change can be made safely in this PR.

**Review by vyasr — CHANGES_REQUESTED** [review]
Back to requesting the change :)

File: python/cudf/cudf/core/multiindex.py
Comment (bdice) [inline python/cudf/cudf/core/multiindex.py:790]: The failures I observed are resolved by #10114.
```suggestion
            result.index = index
```

**Review by vyasr — APPROVED** [review]

**Comment by bdice** [comment]
@gpucibot merge

**Pull request merged**
